### PR TITLE
fix(chartutil): port over enhancements to `helm create` from Helm 2

### PR DESCRIPTION
**Pull request by bacongobbler**
Signed-off-by: Matthew Fisher <matt.fisher@microsoft.com>

### CODE REVIEW

File: pkg/chartutil/create.go [hunk 1]
```diff
@@ -47,10 +47,14 @@ const (
 	DeploymentName = TemplatesDir + sep + "deployment.yaml"
 	// ServiceName is the name of the example service file.
 	ServiceName = TemplatesDir + sep + "service.yaml"
+	// ServiceAccountName is the name of the example serviceaccount file.
+	ServiceAccountName = TemplatesDir + sep + "serviceaccount.yaml"
 	// NotesName is the name of the example NOTES.txt file.
 	NotesName = TemplatesDir + sep + "NOTES.txt"
 	// HelpersName is the name of the example NOTES.txt file.
```
Comment (hickeyma) [inline pkg/chartutil/create.go]: :s/HelpersName is the name of the example NOTES.txt file./HelpersName is the name of the example NOTES.txt file./HelpersName is the name of the example helpers file.

**Review by hickeyma — CHANGES_REQUESTED** [review]
Thanks @bacongobbler for doing the port. Some comments inline and also the following 2:
- Update this with test dir added: https://github.com/helm/helm/blob/dev-v3/cmd/helm/create.go#L39
- Add ServiceAccountName, TestConnectionName, TemplatesTestsDir to: https://github.com/helm/helm/blob/dev-v3/pkg/chartutil/create_test.go#L52

File: cmd/helm/create_test.go [hunk 2]
```diff
@@ -106,8 +106,8 @@ func TestCreateStarterCmd(t *testing.T) {
 		t.Errorf("Wrong API version: %q", c.Metadata.APIVersion)
 	}
 
-	if l := len(c.Templates); l != 6 {
-		t.Errorf("Expected 5 templates, got %d", l)
+	if l := len(c.Templates); l != 8 {
```
[the remark this answers is not in the document]
Comment (bacongobbler) [reply cmd/helm/create_test.go]: Good call. Yes, that is a better way to write this.

File: pkg/chartutil/create.go [hunk 3]
```diff
@@ -47,10 +47,14 @@ const (
...
 	HelpersName = TemplatesDir + sep + "_helpers.tpl"
+	// TestConnectionName is the name of the example test hook.
```
[the remark this answers is not in the document]
Comment (bacongobbler) [reply pkg/chartutil/create.go]: Yes, good call. That way if there are future tests we want to add in there they can re-use `TemplatesTestsDir`.

**Comment by bacongobbler** [comment]
Thank you for catching all those, @hickeyma. All comments addressed!

File: cmd/helm/create.go [hunk 4]
```diff
@@ -42,6 +42,7 @@ something like this:
 	├── values.yaml   # The default values for your templates
 	├── charts/       # Charts that this chart depends on
 	└── templates/    # The template files
+	└── tests/        # The test files
```
Comment (hickeyma) [inline cmd/helm/create.go]: Change 'tests/' --> 'templates/tests'

File: pkg/chartutil/create_test.go [hunk 5]
```diff
@@ -57,6 +57,7 @@ func TestCreate(t *testing.T) {
 		NotesName,
 		ServiceName,
 		TemplatesDir,
+		TemplatesTestsDir,
```
Comment (hickeyma) [inline pkg/chartutil/create_test.go]: Can you also add: ServiceAccountName & TestConnectionName ?

**Review by hickeyma — CHANGES_REQUESTED** [review]
Thanks for update @bacongobbler. Two small things.

File: pkg/chartutil/create.go [hunk 6]
```diff
@@ -210,8 +231,17 @@ spec:
       labels:
         {{- include "<CHARTNAME>.selectorLabels" . | nindent 8 }}
     spec:
+    {{- with .Values.imagePullSecrets }}
+      imagePullSecrets:
+        {{- toYaml . | nindent 8 }}
+    {{- end }}
+      serviceAccountName: {{ template "<CHARTNAME>.serviceAccountName" . }}
```
Comment (unguiculus) [inline pkg/chartutil/create.go]: `include` is used in all other places. Let's be consistent.

```suggestion
      serviceAccountName: {{ include "<CHARTNAME>.serviceAccountName" . }}
```

File: pkg/chartutil/create.go [hunk 7]
```diff
@@ -259,6 +289,16 @@ spec:
     {{- include "<CHARTNAME>.selectorLabels" . | nindent 4 }}
 `
 
+const defaultServiceAccount = `{{- if .Values.serviceAccount.create -}}
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ template "<CHARTNAME>.serviceAccountName" . }}
```
Comment (unguiculus) [inline pkg/chartutil/create.go]:
```suggestion
  name: {{ include "<CHARTNAME>.serviceAccountName" . }}
```

File: pkg/chartutil/create.go [hunk 8]
```diff
@@ -259,6 +289,16 @@ spec:
...
+  labels:
+{{ include "<CHARTNAME>.labels" . | indent 4 }}
```
Comment (unguiculus) [inline pkg/chartutil/create.go]:
```suggestion
    {{ include "<CHARTNAME>.labels" . | nindent 4 }}
```

File: pkg/chartutil/create.go [hunk 9]
```diff
@@ -334,6 +374,34 @@ Selector labels
...
+{{- end -}}
+{{- end -}}
+`
+
+const defaultTestConnection = `apiVersion: v1
+kind: Pod
+metadata:
+  name: "{{ include "<CHARTNAME>.fullname" . }}-test-connection"
+  labels:
+{{ include "<CHARTNAME>.labels" . | indent 4 }}
```
Comment (unguiculus) [inline pkg/chartutil/create.go]:
```suggestion
    {{ include "<CHARTNAME>.labels" . | nindent 4 }}
```

**Review by unguiculus — CHANGES_REQUESTED** [review]
A few remarks for consistency.

File: pkg/chartutil/create_test.go (hunk 5 above)
Comment (bacongobbler) [reply pkg/chartutil/create_test.go]: done

File: cmd/helm/create.go (hunk 4 above)
Comment (bacongobbler) [reply cmd/helm/create.go]: done!

**Review by unguiculus — APPROVED** [review]

**Review by hickeyma — APPROVED** [review]
LGTM, thanks @bacongobbler

File: pkg/chartutil/create.go [hunk 10]
```diff
@@ -149,16 +173,13 @@ const defaultIgnore = `# Patterns to ignore when building packages.
...
+.vscode/
 `
 
 const defaultIngress = `{{- if .Values.ingress.enabled -}}
 {{- $fullName := include "<CHARTNAME>.fullname" . -}}
 {{- $svcPort := .Values.service.port -}}
-{{- if semverCompare ">=1.14-0" .Capabilities.KubeVersion.GitVersion -}}
 apiVersion: networking.k8s.io/v1
-{{- else -}}
-apiVersion: extensions/v1beta1
```
Comment (marckhouzam) [inline pkg/chartutil/create.go]: I'm curious how come this was removed?  Won't v3 need to work with clusters earlier than 1.14?

Comment (unguiculus) [reply pkg/chartutil/create.go]: I would say the change is in line with the official Kubernetes version support policy (https://kubernetes.io/docs/setup/release/version-skew-policy/). Personally, I never liked having conditional logic for API versions in the starter chart. After all, this is not about Helm supporting older Kubernetes versions. It's about creating a starter chart which you will have to adjust to your needs anyways.

Comment (marckhouzam) [reply pkg/chartutil/create.go]: Thanks for the clarification.  I guess I was surprised that this conditional API version logic was added 2 weeks ago and now removed.  But I had missed that this was the charter start code when I skimmed the change.

**Pull request merged**